### PR TITLE
[S 40] Forgot password

### PR DESCRIPTION
**Pull request by nomarchia**
This branch contents three tasks:
[s-26] Email verification
[s-39] Change user password
[s-40] Forgot password

### CODE REVIEW

File: src/main/java/com/greetingcard/dao/jdbc/JdbcCardDao.java [hunk 1]
```diff
@@ -84,11 +84,6 @@
                     "WHERE uc.card_id = :cardId AND uc.user_id = :userId";
     private static final String DELETE_BY_CARD_ID = "DELETE FROM cards WHERE card_id=? and user_id=?";
     private static final String CHANGE_STATUS_OF_CARD_BY_ID = "UPDATE cards SET status_id = ? where card_id = ?";
-    private static final String GET_ALL_CARDS_BY_USER_ID = "SELECT c.card_id ,c.name, c.background_image, c.card_link, c.status_id," +
-            " u.user_id, u.firstName, u.lastName, u.login, u.email " +
-            "FROM users_cards uc JOIN cards c ON (uc.card_id = c.card_id) " +
-            "JOIN users u ON (c.user_id = u.user_id) " +
-            "WHERE uc.user_id = :id ORDER BY c.card_id";
```
Comment (Yuliia-Sh) [inline src/main/java/com/greetingcard/dao/jdbc/JdbcCardDao.java:87]: I think, this is correct one

File: src/main/java/com/greetingcard/web/controller/UserController.java [hunk 2]
```diff
@@ -60,16 +66,59 @@ public UserController(SecurityService securityService) {
...
                 .build();
         log.info("Registration request for user login: {}", user.getLogin());
         securityService.save(user);
+
+        String accessHash = securityService.generateAccessHash(email, AccessHashType.VERIFY_EMAIL);
+        log.debug("Sending letter to verify user's email address: {}", email);
+        String emailMessageBody = "Please confirm your email address by opening this link:\n " +
+                siteUrl + "/email/verify/" + accessHash;
+        emailService.sendMail(email, "Greeting Card: Verify email", emailMessageBody);
+
```
Comment (Yuliia-Sh) [inline src/main/java/com/greetingcard/web/controller/UserController.java:85]: I think this logic should be in service level

**Review by RomanKlishch — APPROVED** [review]

File: src/main/java/com/greetingcard/web/controller/UserController.java [hunk 3]
```diff
@@ -60,16 +66,59 @@ public UserController(SecurityService securityService) {
...
         log.info("Successfully registered: {}", user);
         return ResponseEntity.status(HttpStatus.CREATED).build();
     }
+
+    @PostMapping(value = "user/verification/{accessHash}", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public void verifyEmail(@PathVariable String accessHash) {
```
Comment (Yuliia-Sh) [inline src/main/java/com/greetingcard/web/controller/UserController.java:91]: json or path variable should come?

File: src/main/java/com/greetingcard/web/controller/UserController.java [hunk 4]
```diff
@@ -60,16 +66,59 @@ public UserController(SecurityService securityService) {
...
+        securityService.verifyAccessHash(accessHash, AccessHashType.VERIFY_EMAIL);
+    }
+
+    @PostMapping(value = "user/password", consumes = MediaType.APPLICATION_JSON_VALUE)
+    public void changePassword(@RequestParam String oldPassword, @RequestParam String newPassword, HttpSession session) {
```
Comment (Yuliia-Sh) [inline src/main/java/com/greetingcard/web/controller/UserController.java:96]: I am not sure that with json RequestParam will work

File: src/main/resources/spring/applicationContext.xml [hunk 5]
```diff
@@ -91,6 +95,25 @@
         <constructor-arg ref="congratulationService"/>
     </bean>
 
+    <bean id="mailSender" class="org.springframework.mail.javamail.JavaMailSenderImpl">
+        <property name="host" value="smtp.gmail.com"/>
+        <property name="port" value="587"/>
+        <property name="username" value="nomarchia2@gmail.com"/>
+        <property name="password" value="GOOGLEwork2020"/>
+        <property name="javaMailProperties">
```
Comment (Yuliia-Sh) [inline src/main/resources/spring/applicationContext.xml:103]: here should be properties, not constants

File: src/test/java/com/greetingcard/service/impl/DefaultEmailServiceITest.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+     * change it in userWithRealEmail.xml dataset. Peace!
+     *
+     * To use your own mailbox as sender, you'll need to update email address credentials of
+     * "mailSender" bean in applicationContext.xml. Also, you need to allow unsafe applications
+     * access in your google account settings;
+     */
+    @Test
+    void sendMail() {
+        service.sendMail("nomarchia2@gmail.com","Test Subject","Test message body");
+    }
```
Comment (Yuliia-Sh) [inline src/test/java/com/greetingcard/service/impl/DefaultEmailServiceITest.java:29]: I think mock should be used for such cases

File: pom.xml [hunk 7]
```diff
@@ -125,14 +126,25 @@
...
+        <dependency>
+            <groupId>com.sun.mail</groupId>
+            <artifactId>javax.mail</artifactId>
+            <version>${email.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework</groupId>
+            <artifactId>spring-context-support</artifactId>
+            <version>${spring.version}</version>
+        </dependency>
```
Comment (Yuliia-Sh) [inline pom.xml:139]: what spring-context-support is used for?

File: src/main/java/com/greetingcard/web/controller/UserController.java (hunk 3 above)
Comment (Yuliia-Sh) [reply src/main/java/com/greetingcard/web/controller/UserController.java:91]: did you check that? do you have tests?

**Review by AlexKievDeveloper — APPROVED** [review]

**Review by Yuliia-Sh — APPROVED** [review]

**Pull request merged**